### PR TITLE
updates to einsum

### PR DESCRIPTION
**Pull request by nikhilmishra000**
Resolving https://github.com/tensorflow/tensorflow/issues/4722:
-- Fix bug where output indices are expected to be in alphabetical order
-- Add to docstring

### CODE REVIEW

File: tensorflow/python/ops/special_math_ops.py [hunk 1]
```diff
@@ -94,7 +94,9 @@ def einsum(axes, *inputs):
   """
   A generalized contraction between tensors of arbitrary dimension.
 
-  Like numpy.einsum.
+  Like numpy.einsum, but does not support:
+  -- ellipses (subscripts like 'ij...,jk...->ik...')
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops.py]: Can you turn those apostrophes into backticks? Then it'll render nicely as code. Also, please put `numpy.einsum` into backticks. All docstrings get rendered as markdown.

File: tensorflow/python/ops/special_math_ops.py [hunk 2]
```diff
@@ -105,37 +107,43 @@ def einsum(axes, *inputs):
...
-
+  indices = str(idx_all)
 
   assert len(idx_in) == len(inputs), \
     "Expected %d inputs but only got %d" % (len(idx_in), len(inputs))
 
-  # transpose inputs so axes are in alphabetical order
+  missing_idx = set(idx_out).difference(idx_all)
+  assert not missing_idx, \
+    "Unknown ouput axes: %s" % missing_idx
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops.py]: Can you make this into an exception (ValueError is fine), and add it to the docstring?

**Review by martinwicke — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/special_math_ops.py (hunk 2 above)
Comment (nikhilmishra000) [inline tensorflow/python/ops/special_math_ops.py]: @martinwicke should I do the same for all of the other assertions too?

File: tensorflow/python/ops/special_math_ops.py [hunk 3]
```diff
@@ -94,27 +94,40 @@ def einsum(axes, *inputs):
   """
   A generalized contraction between tensors of arbitrary dimension.
 
-  Like numpy.einsum, but does not support:
-  -- ellipses (subscripts like 'ij...,jk...->ik...')
-  -- subscripts that reduce to scalars
+  Like `numpy.einsum`, but does not support:
+  * Ellipses (subscripts like `ij...,jk...->ik...`)
+  * Subscripts that reduce to scalars
+
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops.py]: Can you add an "Args:" section here, and after the args section, add a "Returns:" section. See other op docstrings for examples.

File: tensorflow/python/ops/special_math_ops.py [hunk 4]
```diff
@@ -94,27 +94,40 @@ def einsum(axes, *inputs):
...
+  Raises a ValueError if:
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops.py]: Can you reformat this to
```
Raises:
  ValueError: <all your text, reindented to flow after ValueError>
```

File: tensorflow/python/ops/special_math_ops_test.py [hunk 5]
```diff
@@ -176,7 +180,7 @@ def test_invalid(self):
 
       try:
         result = tf.einsum(axes, *inputs)
-      except AssertionError as e:
+      except ValueError as e:
         print(e)
       assert result is None, \
         "An exception should have been thrown."
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops_test.py]: Can you rewrite this to use self.assertRaises?

File: tensorflow/python/ops/special_math_ops_test.py [hunk 6]
```diff
@@ -190,7 +194,7 @@ def test_dim_mismatch(self):
       result = None
       try:
         result = tf.einsum(axes, *inputs)
-      except AssertionError:
+      except ValueError:
         pass
       assert result is None, "An exception should have been thrown."
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops_test.py]: This too

**Review by martinwicke — COMMENTED** [review]
Sorry, missed some things earlier.

**Comment by jart** [comment]
Einsum is not a trivial abstraction. Numpy has 900 lines of test code for einsum ([test_einsum.py](https://github.com/numpy/numpy/blob/master/numpy/core/tests/test_einsum.py)) and 100 lines of Pydoc tests ([einsumfunc.py](https://github.com/numpy/numpy/blob/master/numpy/core/einsumfunc.py)). Ideally we would just use their test cases to show that the behavior is identical. Especially if we document the function as being "like numpy.einsum."

Thank you for the work you put into bringing einsum to TensorFlow @nikhilmishra000 and for being responsive in addressing the concerns in #4722. I just can't help but wonder if this function is more appropriate for the contrib package, since it isn't on par with numpy in terms of functionality and test comprehensiveness. What do you think @martinwicke?

See also: #4886

**Comment by martinwicke** [comment]
I think it is fine in core. There's no experimentation with the interface needed, which is usually the concern. If its functionality is not a complete mirror of numpy's we have to document the differences clearly. 

In this case, it should be noted that transpose doesn't work, are there other differences that are not documented yet?

**Comment by nikhilmishra000** [comment]
The docstring has been updated, and I have added some more tests.

Following this PR, I think the outstanding differences from numpy will be that:
* Ellipses will not be recognized. This shouldn't be too hard to implement, but I don't know that I have the bandwidth to get it done in the near future.
* Subscripts where an axis appears more than once (example: `ii->i` computes the trace of a matrix). Adding support for this seems a little bit more involved.

Exceptions are raised for both of these cases.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by nikhilmishra000** [comment]
The previous failure in python3 was my fault (I didn't realize that passing a comparator to `sorted()` had been disallowed). Not too sure what happened on linux gpu, but it seems unrelated?

**Comment by martinwicke** [comment]
Jenkins, test this please!

File: tensorflow/python/ops/special_math_ops.py [hunk 7]
```diff
@@ -94,48 +94,100 @@ def einsum(axes, *inputs):
...
       inputs[i] = array_ops.transpose(input_, permuted)
       idx_in[i] = sorted_idx
 
-  missing_idx = set(idx_out).difference(idx_all)
-  assert not missing_idx, \
-    "Unknown ouput axes: %s" % missing_idx
-
   reduction_idx = []
   shapes = [[dim if dim else -1
              for dim in tensor.get_shape().as_list()]
```
Comment (eaplatanios) [inline tensorflow/python/ops/special_math_ops.py]: I would like to propose one more fix. I think this should be changed from:

```
shapes = [[dim if dim else -1
           for dim in tensor.get_shape().as_list()]
           for tensor in inputs]
```

to:

```
shapes = [[dim if dim is not None else -1
           for dim in array_ops.unpack(array_ops.shape(tensor))]
           for tensor in inputs]
```

in order to allow for dynamic tensor shapes. I run into an issue while using `tf.einsum` and this change fixed it for me.

**Review by yunita15 — COMMENTED** [review]
Good

File: tensorflow/python/ops/special_math_ops.py (hunk 7 above)
Comment (eaplatanios) [inline tensorflow/python/ops/special_math_ops.py]: Sounds good! I made PR #5174 but without any test coverage. I have a paper deadline approaching and I'm not sure about your test conventions, so I thought I'd leave it there for now. This change alone seems to be working fine for me, although I haven't tried using it with any edge cases.

**Pull request merged**